### ISSUE TITLE
[Question] Single forward pass through ChatModule

### ISSUE THREAD

**Issue by caenopy** [mlc-ai/mlc-llm#2354]
## ❓ General Questions

An older [example](https://github.com/CharlieFRuan/anticipation-mlc-llm/blob/48d9332f7dbb174e23ed5aa0ec5cffe4362efe8e/anticipation/sample.py#L93) shows that you could previously run _forward_tokens() on a ChatModule to get logits. The API has since changed, so how would I do this now?

Thanks!

@CharlieFRuan

**Comment by CharlieFRuan** [comment]
Thanks for the issue, will take a look this weekend. A low-level forward/sample API may still be helpful; DebugChat may be a reasonable place to put it in.

**Comment by caenopy** [comment]
Thanks Charlie! Any update on the status of this?

**Comment by MasterJH5574** [comment]
Hi @caenopy, right now we have `debug_chat.py` which may be relevant to your needs. https://github.com/mlc-ai/mlc-llm/blob/main/python/mlc_llm/testing/debug_chat.py. Specifically, the code that runs the forward and gets the logits is at https://github.com/mlc-ai/mlc-llm/blob/a231ae1215bd7e06a2f5eddb4e826cd873e69820/python/mlc_llm/testing/debug_chat.py#L310-L323.

It's possible that you need to customize `debug_chat.py` a bit for your particular needs, but please definitely check it out. Thanks!

**Comment by caenopy** [comment]
Thanks @MasterJH5574! This should offer what I need. I'm running to a separate issue when trying the DebugChat approach with music models. I opened another issue to sort this out, as it seems partially model specific.
